From: Anisse Astier <anisse@astier.eu>
Date: Mon, 23 Apr 2012 12:33:11 +0200
Subject: rt2800: add chipset revision RT5390R support

commit 0586a11b5cc51413240e6688936e2edac9c2918e upstream.

About 70% of the chips with revision RT5390R initialize incorrectly, using
the auxiliary antenna instead of the main one. The net result is that
signal reception is very poor (no AP further than 1M).

This chipset differs from RT5390 and RT5390F by its support of hardware
antenna diversity. Therefore antenna selection should be done
differently, by disabling software features and previously selected
antenna.

This changeset does just that, and makes all RT5390R work properly.

This is based on Ralink's 2012_03_22_RT5572_Linux_STA_v2.6.0.0_DPO
driver.

Signed-off-by: Anisse Astier <anisse@astier.eu>
Signed-off-by: John W. Linville <linville@tuxdriver.com>
---
 drivers/net/wireless/rt2x00/rt2800.h    |    1 +
 drivers/net/wireless/rt2x00/rt2800lib.c |   12 ++++++++++++
 2 files changed, 13 insertions(+)

Index: linux-3.2.46/drivers/net/wireless/rt2x00/rt2800.h
===================================================================
--- linux-3.2.46.orig/drivers/net/wireless/rt2x00/rt2800.h	2013-07-26 19:20:49.000000000 +0000
+++ linux-3.2.46/drivers/net/wireless/rt2x00/rt2800.h	2013-07-26 19:26:04.000000000 +0000
@@ -83,6 +83,7 @@
 #define REV_RT3090E			0x0211
 #define REV_RT3390E			0x0211
 #define REV_RT5390F			0x0502
+#define REV_RT5390R			0x1502
 
 /*
  * Signal information.
Index: linux-3.2.46/drivers/net/wireless/rt2x00/rt2800lib.c
===================================================================
--- linux-3.2.46.orig/drivers/net/wireless/rt2x00/rt2800lib.c	2013-07-26 19:20:49.000000000 +0000
+++ linux-3.2.46/drivers/net/wireless/rt2x00/rt2800lib.c	2013-07-26 19:26:04.000000000 +0000
@@ -3185,6 +3185,13 @@
 			rt2800_register_write(rt2x00dev, GPIO_CTRL_CFG, reg);
 		}
 
+		/* This chip has hardware antenna diversity*/
+		if (rt2x00_rt_rev_gte(rt2x00dev, RT5390, REV_RT5390R)) {
+			rt2800_bbp_write(rt2x00dev, 150, 0); /* Disable Antenna Software OFDM */
+			rt2800_bbp_write(rt2x00dev, 151, 0); /* Disable Antenna Software CCK */
+			rt2800_bbp_write(rt2x00dev, 154, 0); /* Clear previously selected antenna */
+		}
+
 		rt2800_bbp_read(rt2x00dev, 152, &value);
 		if (ant == 0)
 			rt2x00_set_field8(&value, BBP152_RX_DEFAULT_ANT, 1);
@@ -4099,6 +4106,11 @@
 		rt2x00dev->default_ant.rx = ANTENNA_A;
 	}
 
+	if (rt2x00_rt_rev_gte(rt2x00dev, RT5390, REV_RT5390R)) {
+		rt2x00dev->default_ant.tx = ANTENNA_HW_DIVERSITY; /* Unused */
+		rt2x00dev->default_ant.rx = ANTENNA_HW_DIVERSITY; /* Unused */
+	}
+
 	/*
 	 * Determine external LNA informations.
 	 */
